### ISSUE TITLE
masked_index_add

### ISSUE THREAD

**Issue by RaulPPelaez** [pytorch/pytorch#122092]
### 🚀 The feature, motivation and pitch

We train graph neural networks for molecules, in which the number of edges in the graph changes depending on the changing locations of the nodes. 
CUDA graphs are important for our workloads so we try to maintain static shapes.

For that I usually establish a maximum number of possible edges and mask unused ones. I often find myself implementing a construct similar to this:
```python
import torch


def static_shape_operation(source, index, values):
    # index = index[index >= 0] # This line would make the shape of index dynamic
    # I replace masked elements by a phantom value
    index = index.masked_fill(index < 0, n_sources)
    # Append a phantom value
    source = torch.cat((source, torch.zeros(1, dtype=torch.float32)))
    # All masked elements are summed to the phantom value
    result = source.index_add(dim=0, index=index, source=values)
    # Remove the phantom value
    result = result[:-1]
    return result


n_sources = 4
# Some arbitrary per-node values
source = torch.zeros(n_sources, dtype=torch.float32)

# Some arbitrary per-edge indexes and values, -1 means the edge is masked
index = torch.tensor([0, 1, 2, 3, -1, -1, -1])
values = torch.tensor([3, 2, 3, 8, -1, -1, -1], dtype=torch.float32)

result = static_shape_operation(source, index, values)
assert result.shape[0] == n_sources
```
Given that index_add can be implemented using atomic operations and the number of masked edges is arbitrary (and possibly high) this is a large price to pay for static shapes.

Instead, I would like to implement the function as:

```python
def static_shape_operation(source, index, values):
    mask = index > 0
    return source.masked_index_add(dim=0, index=index, source=values, mask=mask)
```
Similarly to what can be done for masked_scatter.

**Comment by ezyang** [comment]
Maybe this can be conveniently implemented with torch.compile. We should have all the pieces, try writing the straightforward reference implementation and toss a torch.compile on it.

**Comment by isuruf** [comment]
With https://github.com/pytorch/pytorch/pull/116491 you can do `torch._unsafe_masked_index_put_accumulate(source, index >= 0, [index], values)`
